### PR TITLE
feat: Scotch mapping support

### PR DESCRIPTION
**Pull request by DanGiles**

### CODE REVIEW

**Comment by MatthieuSchaller** [comment]
Could you run the code formatting script to help with the diffing?

**Comment by DanGiles** [comment]
No problem. Thought I did [this](https://github.com/SWIFTSIM/SWIFT/pull/38/commits/ca0f0184536de5dd549e79bf31c7cc1a6ce07407) yesterday however before creating the PR. Will go again and update.

File: src/const.h
Comment (pwdraper) [inline src/const.h:1]: These whitespace changes here and elsewhere are not needed. Please revert them to clean up the merge.

File: src/partition.c [hunk 1]
```diff
@@ -63,6 +67,9 @@
 #include "threadpool.h"
 #include "tools.h"
 
+#define IDX_MAX INT32_MAX
```
Comment (pwdraper) [inline src/partition.c:70]: These defines are not going to work with METIS builds. Not sure if scotch has a 32/64 bit build choice like
METIS, but these are defined as INT32_ and INT64_ as needed in `metis.h`. At least wrap them in an '#ifndef'.

File: src/partition.c [hunk 2]
```diff
@@ -1349,13 +1359,348 @@ static void pick_metis(int nodeID, struct space *s, int nregions,
 }
 #endif
 
-#if defined(WITH_MPI) && (defined(HAVE_METIS) || defined(HAVE_PARMETIS))
+#if defined(WITH_MPI) && defined(HAVE_SCOTCH)
+/**
+ * @brief Fill the adjncy array defining the graph of cells in a space.
+ *
+ * See the ParMETIS and METIS manuals if you want to understand this
```
Comment (pwdraper) [inline src/partition.c:1366]: I remember that this is all modelled on METIS for compatibility, but is it true that SCOTCH points you at the METIS
documentation?

Comment (pwdraper) [reply src/partition.c:1366]: BTW, is it SCOTCH, Scotch or scotch? Always nice to be consistent with the proper name.

File: src/partition.c [hunk 3]
```diff
@@ -1349,13 +1359,348 @@ static void pick_metis(int nodeID, struct space *s, int nregions,
...
+    }
+
+    if (SCOTCH_graphSave(&graph, graph_file) != 0) {
+      printf("Error: Cannot save Scotch Graph.\n");
+    }
+    fclose(graph_file);
+    // #endif
+    /* Read in architecture graph. */
+    SCOTCH_Arch archdat;
+    /* Load the architecture graph in .tgt format */
```
Comment (pwdraper) [inline src/partition.c:1635]: Structurally this part looks odd to me. I think we should really be opening this file once during the 
initialisation and then caching whatever is needed for the constructions here.

File: src/const.h
Comment (DanGiles) [reply src/const.h:1]: Will do. FYI: These were generated using [format.sh](https://github.com/SWIFTSIM/SWIFT/blob/master/format.sh) so the formatting scripts my need updating.

File: src/partition.c [hunk 4]
```diff
@@ -1387,7 +1732,7 @@ void partition_gather_weights(void *map_data, int num_elements,
 
   double *weights_e = mydata->weights_e;
   double *weights_v = mydata->weights_v;
-  idx_t *inds = mydata->inds;
+  int *inds = mydata->inds;
```
Comment (pwdraper) [inline src/partition.c:1735]: As with the over `IDX` related types these can be 64 bit, so we'll need to take more care about this.

Comment (pwdraper) [reply src/partition.c:1735]: Just checking I see there is a  type `SCOTCH_Idx` and define `SCOTCH_NUMMAX`. Probably suitable for sorting
out a unified type.

File: src/partition.c (hunk 1 above)
Comment (pwdraper) [reply src/partition.c:70]: The `repartition_name[]` array needs a final entry for the SCOTCH repartitioning description.

Comment (pwdraper) [reply src/partition.c:70]: Just seeing `null` reported at present.

File: src/partition.c [hunk 5]
```diff
@@ -1349,13 +1359,348 @@ static void pick_metis(int nodeID, struct space *s, int nregions,
...
+    if (SCOTCH_stratGraphMapBuild(&stradat, flagval, num_vertices, 0.05) != 0)
+      error("Error setting the Scotch mapping strategy.");
+
+    /* Map the computation graph to the architecture graph */
+    if (SCOTCH_graphMap(&graph, &archdat, &stradat, regionid) != 0)
+      error("Error Scotch mapping failed.");
+    // #ifdef SWIFT_DEBUG_CHECKS
+    SCOTCH_Mapping mappptr;
+    SCOTCH_graphMapInit(&graph, &mappptr, &archdat, regionid);
+    FILE *map_stats = fopen("map_stats.out", "w");
```
Comment (pwdraper) [inline src/partition.c:1659]: Probably want to make this a debugging option as well.

File: src/partition.c [hunk 6]
```diff
@@ -1349,13 +1359,348 @@ static void pick_metis(int nodeID, struct space *s, int nregions,
...
+    SCOTCH_graphInit(&graph);
+
+    if (SCOTCH_graphBuild(&graph, baseval, vertnbr, verttab, NULL, velotab,
+                          NULL, edgenbr, edgetab, edlotab) != 0) {
+      error("Error: Cannot build Scotch Graph.\n");
+    }
+    // #ifdef SWIFT_DEBUG_CHECKS
+    SCOTCH_graphCheck(&graph);
+    static int partition_count = 0;
+    char fname[200];
```
Comment (pwdraper) [inline src/partition.c:1621]: Also debugging here as well.

File: src/partition.c (hunk 4 above)
Comment (DanGiles) [reply src/partition.c]: Reverted back to `idx_t` which is now defined for Scotch [case](https://github.com/UCL/swiftsim/blob/cb06b0e5c3d8457c474d0084d973f437d29b20d8/src/partition.c#L71-L76)

File: configure.ac [hunk 7]
```diff
@@ -778,6 +778,41 @@ AC_CHECK_LIB(pthread, posix_fallocate,
...
+
+SCOTCH_LIBS=""
+if test "x$with_scotch" != "xno"; then
+
+# Check if we have SCOTCH.
+   if test "x$with_scotch" != "xyes" -a "x$with_scotch" != "x"; then
+      SCOTCH_LIBS="-L$with_scotch/lib -lscotch -lscotcherr"
+      SCOTCH_INCS="-I$with_scotch/include"
+   else
+      SCOTCH_LIBS="-lscotch"
```
Comment (pwdraper) [inline configure.ac:799]: Need -lscotcherr here as well.

**Comment by MatthieuSchaller** [comment]
Are there updates on the points Peter made about the code?

**Pull request closed without merging**